### ISSUE TITLE
New Rule: no-important

### ISSUE THREAD

**Issue by aryan02420** [eslint/css#20]
### Rule details

Flag the use of !important

### What type of rule is this?

Warns about a potential problem

### Example code

```css
.really-red {
    color: red !important;
}
```

### Participation

- [x] I am willing to submit a pull request to implement this rule.

### Additional comments

Needing `!important` indicates there may be a larger underlying issue.

**Comment by jeddy3** [comment]
As with https://github.com/eslint/css/issues/19, a [`no-restricted-syntax`](https://eslint.org/docs/latest/rules/no-restricted-syntax) rule could cover this use case:

```json
{
    "rules": {
        "no-restricted-syntax": [
            "error",
            {
                "selector": "Declaration[important!=true]",
                "message": "!important annotations are not allowed."
            },
        ]
    }
}
```

When using [cascade layers](https://drafts.csswg.org/css-cascade-5/#layering), the `!important` annotation [is useful](https://css-tricks.com/css-cascade-layers/#aa-important-layers):

> Once we start using cascade layers, we will need to be much more cautious and intentional about how we use `!important`. It’s no longer a quick way to jump to the top of the priorities — but an integrated part of our cascade layering; a way for lower layers to _insist_ that some of their styles are essential.

Fewer people will likely want to restrict this syntax as they adopt cascade layers.

**Comment by nzakas** [comment]
Yeah, I think we want to hold off on adding rules that simply disallow syntax because it can easily be covered by `no-restricted-syntax`, but leaving open for now to see if others feel differently.

**Comment by yannbertrand** [comment]
I was excited about that one but reading the comments makes me doubt now 😀. I believe @jeddy3's view on these is super interesting, I personally have 0 background on CSS linting so following your advice seems like the best idea to me!

**Comment by nzakas** [comment]
@jeddy3 is this true for keyframe `!important` as well? I noticed Stylelint has two different rules for `!important`:

https://stylelint.io/user-guide/rules/keyframe-declaration-no-important/
https://stylelint.io/user-guide/rules/declaration-no-important/

**Comment by jeddy3** [comment]
The [`declaration-no-important`](https://stylelint.io/user-guide/rules/declaration-no-important/) rule for restricting valid syntax, whereas the [`keyframe-declaration-no-important`](https://stylelint.io/user-guide/rules/keyframe-declaration-no-important/) one flags [invalid `!important` within keyframes](https://drafts.csswg.org/css-animations/#keyframes): 

> The [`<declaration-list>`](https://www.w3.org/TR/css-syntax-3/#typedef-declaration-list) inside of [`<keyframe-block>`](https://drafts.csswg.org/css-animations/#typedef-keyframe-block) accepts any CSS property except those defined in this specification, but does accept the [animation-timing-function](https://drafts.csswg.org/css-animations/#propdef-animation-timing-function) property and interprets it specially. None of the properties interact with the cascade (so _using !important on them is invalid and will cause the property to be ignored_).

We added the [`declaration-no-important`](https://stylelint.io/user-guide/rules/declaration-no-important/) rule to Stylelint at a time when using `!important` was considered universally bad. This no longer appears to be the case, especially when [using cascade layers](https://github.com/eslint/css/issues/20#issuecomment-2504336404).

I'd suggest:

1. Repurposing https://github.com/eslint/css/pull/23 to disallow only _invalid_ `!important` annotations and turning it on in your recommended config.
2. Adding a [`no-restricted-syntax`](https://eslint.org/docs/latest/rules/no-restricted-syntax) rule, using valid `!important` annotations as an example, and not turning on the rule in your recommended config.

You could either limit the repurposed https://github.com/eslint/css/pull/23 to keyframe blocks (as `no-invalid-important`) or also flag other invalid annotations (as `no-invalid-annotation`), e.g.:

```css
a {
  color: red !foo;
}
```

(You may want to allow [`<dashed-ident>`](https://drafts.csswg.org/css-values/#dashed-idents) annotations, e.g. `!--foo`, as [author-controlled syntax seems to be going that way](https://drafts.csswg.org/css-values/#example-32974307).)

**Comment by mdjermanovic** [comment]
> 2\. Adding a `no-restricted-syntax` rule, using valid `!important` annotations as an example, and not turning it on in your recommended config.

The core [`no-restricted-syntax`](https://eslint.org/docs/latest/rules/no-restricted-syntax) rule can already be used with any language, including css.

**Comment by jeddy3** [comment]
Wow, mind 🤯. Thank you for pointing that out!

Using it, I was also able to replicate nearly all of Stylelint's [`*-(dis)allowed-list`](https://stylelint.io/user-guide/rules#allowed-disallowed--required) rules.

From simple ones like [`function-disallowed-list`](https://stylelint.io/user-guide/rules/function-disallowed-list), e.g. disallowing the `var()` and `rgb()` functions:

```js
{
  selector: "Function[name=/\\b(var|rgb)\\b/]",
  message: "var() and rgb() functions are not allowed",
}
```

To pair matching ones like [`declaration-property-unit-disallowed-list`](https://stylelint.io/user-guide/rules/declaration-property-unit-disallowed-list), e.g. disallow the `s` unit for the `animation` and `animation-duration` properties:

```js
{
  selector: "Declaration[property=/^animation/] Dimension[unit=s]",
  message: "Use `ms` instead of `s` for animation durations"
},
```

Because ESQuery supports the `:has()` selector, I could also replicate (and improve on the) rules that would require inspecting the attributes of child descendant nodes like [`media-feature-name-unit-allowed-list`](https://stylelint.io/user-guide/rules/media-feature-name-unit-allowed-list), e.g. enforcing `rem` units for `width` features:

```js
{
  selector: "FeatureRange:has(Identifier[name=width]) Dimension[unit!=rem]",
  message: "Use `rem` units for the `width` feature"
},
```

It would catch the use of `px` units in both ranged media and container queries:

```css
@media screen and (10px < width <= 768px) {}

div {
  container-type: inline-size;
  div {
    @container (width > 30px) {}
  }
}
```

**Comment by esycat** [comment]
I had `!important` forbidden for years and would prefer a dedicated rule even if the same can be achieved via a generic one.

**Comment by nzakas** [comment]
Thanks for the feedback. What we're really looking for here is whether or not a `no-important` rule makes sense when using layers, and if so, what (if any) modifications should there be to the behavior?

**Comment by nzakas** [comment]
After doing a bunch of reading and research, I've come to the conclusion that it makes sense to have a `no-important` rule to forbid `!important` throughout a stylesheet. I've not come across a good argument for allowing `!important` when using layers. If one should pop in the future we can always add an option to allow for it.
